### ISSUE TITLE
eksctl apply: cluster reconciliation

### ISSUE THREAD

**Issue by michaelbeaumont** [weaveworks/eksctl#2774]
**Why do you want this feature?**
This is the umbrella issue concerning `eksctl apply`, which would support `eksctl apply -f config.yaml` to reconcile the current cluster with the given config (initially partially). Also sometimes known as `eksctl update`.

### Related issues:
#1497 
#462 (previous umbrella issue, closed because the history is confusing)
#20

**Comment by Callisto13** [comment]
🎉  

This issue is going to track the delivery of a proposal for how we will implement gitops-style reconciliation with eksctl.

Deliverable is a proposal in `docs/` which details:
  1. The end state of `eksctl apply` (ie. ignoring everything else which currently exists, what would it look like if we could `apply` today?)
  1. What the rest of the `eksctl` UX looks like in a world with `apply` (ie. which flags, subcommands, etc would survive?)
  1. An implementation strategy: how we adapt the code to get from now to `apply` (design diagrams may be handy)
  1. A consideration of the risks and anything which may stop us from truly representing a cluster in config


Doing gitops reconciliation has been the goal of this project for a very long time, but the rollout was so gradual that the Plan was eventually forgotten as people moved off the project. We need to move incrementally, but let's aim to not be too incremental so that we forget the big picture.

Please add comments on what you would like to see in a proposal doc for this work, which questions you would like answered, etc

**Comment by aclevername** [comment]
>1. The end state of eksctl apply (ie. ignoring everything else which currently exists, what would it look like if we could apply today?)

I think sticking to a relatively consistent state to the existing `kubectl apply` (so no sub commands afterwards) would be ideal, perhaps with some slight modifications (opinions/thoughts welcome):
- Keeping the dry run by default behaviour: using the existing `--approve` flag to state when you should actually create the resources. This differs from apply, but given the nature of eksctl and cost associated with creating clusters having a dry run by default feels nice. By default it could just show you the differ of the current world vs whats defined in the cluster config.
- Having the `--prune` functionality by default. Given the large majority of what we create is cloudformation stacks, finding all stacks that have been created that are no longer specified in the config isn't too complicated (regex name and/or tags). Might be some edge case resources where we don't use cloudformation.


> 2. What the rest of the `eksctl` UX looks like in a world with `apply` (ie. which flags, subcommands, etc would survive?)

I think this is an interesting one. Removing existing subcommands would presumably mean we are doing a breaking change, so a new major (1.0 🎉 ?)

I think there is some value in keeping some of the imperative command, for demonstrating eksctl being able to run `eksctl create cluster --name jk` and have a cluster 20 minutes later is pretty amazing and provides a relatively easy introduction to eksctl. Similar to how `kubectl create deployment nginx --image=nginx ` exist currently. It might be that some of the more advanced ones could be removed in order to push users into using config files.

This might provide a good opportunity for us to remove some of the existing behaviour of deleting nodegroups from a config file, which currently has a view variations:
```      
--include strings             nodegroups to include (list of globs), e.g.: 'ng-team-?,prod-*'
--exclude strings             nodegroups to exclude (list of globs), e.g.: 'ng-team-?,prod-*'
--only-missing                Only delete nodegroups that are not defined in the given config file
``

**Comment by Callisto13** [comment]
excellent point!

**Comment by artem-nefedov** [comment]
Absolutely would love to see this.
I would argue that this is the most important feature moving forward, because the main point of `eksctl` is that it **simplifies** the management of EKS clusters.
However, we had to implement multiple workarounds and hacks already just because `eksctl upgrade cluster -f config.yaml` or `eksctl upgrade nodegroup -f config.yaml` commands doesn't want to honor changes made to config manifest, which kinda undermines the whole simplification part.

**Comment by michaelbeaumont** [comment]
To clarify for 2.), I don't see the alteration of any existing commands as in scope for this issue.

**Comment by Callisto13** [comment]
👍 totally. I didn't mean "alteration", more "termination" 😈

**Comment by michaelbeaumont** [comment]
I think 3) is sort of dependent on 1) being "finalized", right? That is, I don't think a proposal for the behavior of `apply` should need to answer implementation questions, although we should consider 4).

**Comment by Callisto13** [comment]
🤔  Isn't explaining how we will do the thing one of the core parts of a design proposal? It's not like we can start work without it/a bunch of tickets

**Comment by michaelbeaumont** [comment]
I would have thought the first step/proposal is something users might engage with. It would answer the question "how does `eksctl apply` behave with this config?".

**Comment by Callisto13** [comment]
A complete proposal should cover all stages, all questions, and all risks, but we could get away with not providing them all at once? The danger is that sometimes risks and questions discovered later can end up influencing the behaviour.

Either way I am not bothered if we take each step separately so long as [all information is covered](https://github.com/Callisto13/eksctl/blob/cb-contrib-docs-changes/CONTRIBUTING.md#proposals) before we start work.

**Comment by bitva77** [comment]
Just want to say that if this feature means I can stop gitops'ing `eksctl` with bash scripts, this would be such a win.

/going back into my hole now

**Comment by nickjj** [comment]
I'd really like to see this too.

**Use case**: I wanted to attach a new iam addon policy to an existing cluster. I ended up doing:

1. Update my cluster config file with the new addon
2. Create a new node group with a new name (arbitrarily renamed it by adding -abc to the end of the name)
3. Delete the old node group

This worked but it was a pretty big ordeal that took over 8 minutes while waiting for AWS and also performing destructive actions. This all boils down to adding a policy to a group of nodes right? Would be great if you could run 1 command have the cloudformation templates apply them to the existing nodes.

**Comment by kahirokunn** [comment]
Is there any progress?

**Comment by steffakasid** [comment]
This would really be the killer feature to be able to use eksctl e.g. in pipelines to roll-out clusters & cluster updates.

**Comment by EliMor** [comment]
Personal opinion, without this feature I'm not comfortable using `eksctl` in production. Very cool project though, I hope it keeps going.

**Comment by bitva77** [comment]
Things have improved in this area.  We use eksctl in Production just fine.  I mean, how often are you changing EKS control plane configuration?  Like never.  Node groups are the only thing we change and I'm not sure I want node groups to be deleted automatically yet...especially in Production.

**Comment by Skarlso** [comment]
>Very cool project though, I hope it keeps going.

That's a bit passive-aggressive. :) It was created in 2018 and has been going strong for a while now, without any indication of stopping. :) And it just gets better and better.

While apply IS a killer feature, you can combine eksctl with flux easily to support this workflow by eksctl creating the necessary files and flux managing them. `eksctl` even sports a flux integration itself. Or you can use `eksctl` together with terraform to achieve a declarative description of your infrastructure.

We might eventually support `apply`, but there are many features that have precedence. Like, as supporting Karpenter out of the box, for example, allowing users to explore and use Karpenter in a friendly and easy way. :)

**Comment by artem-nefedov** [comment]
There are a lot of things that can be safely changed in-place, but you have to perform a separate command for each of them, which is extremely annoying and un-declarative.

These include (but not limited to):

- AWS resource tags
- IAM serviceaccounts
- public cluster API endpoint CIDR ranges

If `eksctl apply` could handle just those "safe" cases, that would already be wonderful.

**Comment by Skarlso** [comment]
And we graciously accept community contributions for these features and will happily help get the PR ready and merged! :)

**Comment by rverma-nsl** [comment]
One thing I want to suggest is that eksctl apply doesn't need to be a functionality within the eksctl only, even if there is a GitHub action which can simulate reconciliation, that is also good enough.

All we need to do is just place the right group of commands and hooks.
e.g. Even if we don't have inplace node reconcillation, this can be done using
1. Mandate an unique node group name (may include version or date in the name)
2. Create and remove orphan node groups

However the most challenging part of the scripts would be recovering from errors and achieve Idempotency. If someone is already  have these in bits and pieces as a community we can convert this to a GitHub action.

**Comment by mhemken-vts** [comment]
Hold on... there's no such thing as `apply`? Am I having a case of the Mandela Effect? I swear I've used `eksctl apply` before.

How was this not a thing from the beginning?

**Comment by mhemken-vts** [comment]
> We might eventually support `apply`, but there are many features that have precedence. Like, as supporting Karpenter out of the box, for example, allowing users to explore and use Karpenter in a friendly and easy way. :)

How does Karpenter have precedence over `apply`? Isn't the whole point of declarative infra that you can do idempotent applys repeatedly? I thought that was the whole point of this tool.

For those using this tool in production, what is the normal development cycle? Do you just create a new cluster when the old one's configuration inevitably becomes outdated?

**Comment by bitva77** [comment]
I think you're confusing `eksctl` with `terraform`

**Comment by matti** [comment]
@mhemken-vts I'm creating a new cluster and new nodegroups when I need to change something. It's horrible, I know.

However, apply would be great only for cluster.

I honestly think its better to create new nodegroups and then delete old ones as if your change is problematic, your old nodesgroups are still working.

**Comment by alexmaccabi** [comment]
oh wow, reading all the comments from the past and its still not implemented ???

**Comment by MartinEmrich** [comment]
Yes, comments do not equal code. As a long-term extensive eksctl user, I can understand that it is definitively not trivial to implement a do-it-all declarative approach through the imperative AWS API.

But you can do every change imperatively (upgrade Control Plane, Upgrade Addons, CRUD on IAM Service Accounts, replace nodegroups, ....). We have the same EKS cluster running since about EKS 1.18 and it is still in tip-top shape.
 
I see no reason to abandon a cluster in need of a change.
